dataset/groupby.cpp

```diff
@@ -140,20 +141,17 @@ template <void (*Func)(Variable &, const Variable &)>
 // requiring the introduction of a wrapping struct to aid compiler
 // resolution.
 struct wrap {
+  template <variable::FillValue fill>
   static constexpr auto reduce_idempotent =
       [](auto &&out, const auto &data_container,
          const GroupByGrouping::group &group, const Dim reductionDim,
          const Variable &mask) {
-        bool first = true;
+        copy(special_like(out, fill), out);
```

Comment:
Add an out arg to `special_like` or (maybe better) add a `fill` function instead? Would avoid doing two passes over the data and creating a temporary.

Reply:
I would have said "premature optimization", but actually I managed to refactor nicely, at the same time also enabling a working `groupby().sum()` for `bool`, which previously failed since it created output of type `bool`. See update.